overlay/files/system/oem/05_network.yaml

```diff
@@ -13,6 +13,13 @@ stages:
       commands:
         - rm /etc/resolv.conf
         - ln -s /run/systemd/resolve/resolv.conf /etc/resolv.conf
+  initramfs.after:
+    - name: "Restart network"
+      commands:
+        - systemctl daemon-reload
+        - systemctl restart systemd-networkd
+        - |
+          ip link | awk -F: '$0 !~ "NO-CARRIER|lo|vir|wl|usb|^[^0-9]"{print $2;getline}'| while read line; do /sbin/dhclient -4 $line; done
```

Comment:
this approach is really invasive and might clash with static IPs setups. I wonder, isn't equivalent doing the same when dumping your network configuration in the first place?

Reply:
The `dhclient -4` command has no effect on statically configured interfaces. It will only trigger a DHCP refresh on interfaces that are supposed to get DHCP addresses. The goal of the `awk` command is to find the interfaces that wired up and can actually get their DHCP address, skipping interfaces that not connected and could cause DHCP delays for other interfaces. We need to have the DHCP address in place before stylus-agent starts flashing the local HDD, hence it is extremely important that we do our best to ensure the addresses are in place.

Reply:
indeed, I'm just asking if this can't be provided with the cloud config supplied during installation? what's the catch I'm missing here?

I say this also because I'm still not really convinced by this solution. relies on too many moving pieces that might not _always_ do what you expect here: 
- on `awk` - which version do we have ? do we have `awk`? we never explictly relied on it until now
- a certain implementation on `dhclient` present (did you checked that the same flags works for e.g. in the dhclient busybox? or in the opensuse/ubuntu version of dhclient?
- assumes that there is a configuration going on between the stages, which is the specific case, and not a general one

Reply:
The problem is that the NICs in Equinix Metal are not predictable. Some servers have only 2 onboard NICs, others have both the onboard NICs and an additional NIC card. When the additional NIC card is present, that one is wired up and the onboard NICs are not. Hence there is no way to know upfront which `eth/eno/enp` interface is the one that needs to be used for DHCP in that server. Hence the iPXE bootfile uses `ip=dhcp` (without specifying a specific NIC) and the cloud-config uses a script to dynamically query the correct NIC config from the Equinux Metal metadata API and build the systemd-networkd networking files for the next boot (registration phase).

It's just the installation phase that we need to get through. What could be done is to gate the `ip link | awk | dhclient` command behind an `if: "[ -f /run/cos/live_mode ]"` statement, so that it only fires during the installation phase, which is the only time it's really needed.

I'm using OpenSUSE since that's in Stylus, but I imagine that since we use systemd-networkd everywhere for Kairos, that `dhclient` will thus be present everywhere too?

Reply:
I tried using just `/sbin/dhclient -4` to avoid the `awk` statement but this suffered from the same DHCP delays and caused kairos to have already installed the image onto the HDD before the system had an IP address. Hence it couldn't retrieve the metadata from the `config_url` parameter before flashing and the necessary customization from the cloud-config does not happen. Since that customization is responsible for building the network bonding info, upon the next boot, the system has no working network stack.

Reply:
@kairos-io/maintainers  thoughts?

Reply:
>Is this happening only on the installation? That sounds less invasive, let's do this.

Yes the problem only occurs on the installation. I will test this variant and update the PR if it works:
```
  initramfs.after:
    - name: "Restart network"
      commands:
        - systemctl daemon-reload
        - systemctl restart systemd-networkd
    - name: "Force DHCP refresh before installation"
      if: "[ -f /run/cos/live_mode ]"
      commands:
        - |
          ip link | awk -F: '$0 !~ "NO-CARRIER|lo|vir|wl|usb|^[^0-9]"{print $2;getline}'| while read line; do /sbin/dhclient -4 $line; done
```

This assumes Kairos will execute the "Force DHCP refresh before installation" after the "Restart network" step, if this is not guaranteed, please let me know how I can ensure this.

Reply:
>are you sure this is still the case with 1.6.0? this looks a timing issue

Yes this is on 1.6.0, no change vs 1.5.0 as this issue occurs when the Login Prompts targets is reached, so all the way at the end of the first startup cycle, just before kairos starts installing the OS onto the disk.

Reply:
Execution order could not be guaranteed, so I moved the command back into the "Restart network" step but gated it with a bash `if` statement. This works correctly.
Also changed `initramfs.after` to `boot.after`, since I noticed that `initramfs.after` was too early for subsequent boots (i.e. registration mode).